File: src/gx/commands/handoff.py
"""gx handoff: Branch summary generator for PRs, Slack, or standups."""

from __future__ import annotations

import platform
import subprocess
from dataclasses import dataclass
from typing import Optional

import typer

from gx.utils.display import console, print_error, print_info, print_success, print_warning
from gx.utils.git import (
    GitError,
    branch_exists,
    ensure_git_repo,
    get_current_branch,
    get_head_branch,
    run_git,
)
from gx.utils.stack import get_parent


@dataclass
class HandoffData:
    branch: str
    base: str
    commits: list[tuple[str, str]]  # (short_hash, message)
    stat_summary: str
    files: list[str]


def _gather_data(base: str) -> HandoffData:
    current = get_current_branch()

    # Commits
    log_output = run_git(["log", "--oneline", "--no-decorate", f"{base}..HEAD"], check=False)
    commits: list[tuple[str, str]] = []
    for line in log_output.strip().splitlines():
        if line:
            parts = line.split(" ", 1)
            if len(parts) == 2:
                commits.append((parts[0], parts[1]))

    # File stats summary
    stat_output = run_git(["diff", "--stat", f"{base}...HEAD"], check=False)
    stat_summary = ""
    if stat_output:
        stat_lines = stat_output.strip().splitlines()
        if stat_lines:
            stat_summary = stat_lines[-1].strip()

    # Changed files
    files_output = run_git(["diff", "--name-only", f"{base}...HEAD"], check=False)
    files = [f for f in files_output.strip().splitlines() if f]

    return HandoffData(
        branch=current,
        base=base,
        commits=commits,
        stat_summary=stat_summary,
        files=files,
    )


def _format_plain(data: HandoffData, is_stacked: bool) -> str:
    if is_stacked:
        header = f"{data.branch} (on {data.base})"
    else:
        header = f"{data.branch} (vs {data.base})"

    lines = [header, ""]
    lines.append(f"Commits ({len(data.commits)}):")
    for short_hash, message in data.commits:
        lines.append(f"  {short_hash}  {message}")
    lines.append("")
    lines.append(data.stat_summary)
    lines.append("")
    lines.append("Files:")
    for f in data.files:
        lines.append(f"  {f}")

    return "\n".join(lines)


def _format_markdown(data: HandoffData) -> str:
    lines = [
        f"## {data.branch}",
        f"**Base:** {data.base} "
        f"· **{len(data.commits)} commit{'s' if len(data.commits) != 1 else ''}** "
        f"· {data.stat_summary}",
        "",
        "### Commits",
    ]
    for short_hash, message in data.commits:
        lines.append(f"- `{short_hash}` {message}")
    lines.append("")
    lines.append("### Files Changed")
    for f in data.files:
        lines.append(f"- `{f}`")

    return "\n".join(lines)


def _copy_to_clipboard(text: str) -> bool:
    system = platform.system()
    try:
        if system == "Darwin":
            subprocess.run(["pbcopy"], input=text.encode(), check=True)
        elif system == "Linux":
            try:
                subprocess.run(
                    ["xclip", "-selection", "clipboard"],
                    input=text.encode(), check=True,
                )
            except FileNotFoundError:
                subprocess.run(
                    ["xsel", "--clipboard", "--input"],
                    input=text.encode(), check=True,
                )
        elif system == "Windows":
            subprocess.run(["clip"], input=text.encode(), check=True)
        else:
            return False
        return True
    except (subprocess.CalledProcessError, FileNotFoundError):
        return False


def handoff(
    against: Optional[str] = typer.Option(None, "--against", help="Compare against a specific branch."),
    copy: bool = typer.Option(False, "--copy", "-c", help="Copy output to clipboard."),
    markdown: bool = typer.Option(False, "--markdown", "--md", help="Output in markdown format."),
) -> None:
    """Generate a branch summary for PRs, Slack, or standups."""
    try:
        ensure_git_repo()
    except GitError as e:
        print_error(str(e))
        raise typer.Exit(1)

    try:
        current = get_current_branch()
    except GitError as e:
        print_error(str(e))
        raise typer.Exit(1)

    # Determine base
    is_stacked = False
    if against:
        if not branch_exists(against):
            print_error(f"Branch '{against}' does not exist.")
            raise typer.Exit(1)
        base = against
    else:
        parent = get_parent(current)
        if parent:
            base = parent
            is_stacked = True
        else:
            base = get_head_branch()

    # Gather
    data = _gather_data(base)

    if not data.commits:
        print_info(f"No changes between {current} and {base}.")
        raise typer.Exit(0)

    # Format
    if markdown:
        output = _format_markdown(data)
    else:
        output = _format_plain(data, is_stacked)

    console.print(output)

    # Clipboard
    if copy:
        if _copy_to_clipboard(output):
            console.print()
            print_success("Copied to clipboard.")
        else:
            console.print()
            print_warning("Could not copy to clipboard.")

File: cmd/handoff.go
package cmd

import (
	"fmt"
	"strings"

	"github.com/mubbie/gx-cli/internal/git"
	"github.com/mubbie/gx-cli/internal/stack"
	"github.com/mubbie/gx-cli/internal/ui"
	"github.com/spf13/cobra"
)

type handoffCommit struct{ hash, msg string }

func init() {
	cmd := &cobra.Command{
		Use:   "handoff",
		Short: "Generate a branch summary for PRs, Slack, or standups",
		RunE:  runHandoff,
	}
	cmd.Flags().String("against", "", "Compare against a specific branch")
	cmd.Flags().BoolP("copy", "c", false, "Copy output to clipboard")
	cmd.Flags().Bool("markdown", false, "Output in markdown format")
	cmd.Flags().Bool("md", false, "Output in markdown format")
	rootCmd.AddCommand(cmd)
}

func runHandoff(cmd *cobra.Command, args []string) error {
	if err := git.EnsureRepo(); err != nil {
		ui.PrintError(err.Error())
		return nil
	}

	current, err := git.CurrentBranch()
	if err != nil {
		ui.PrintError(err.Error())
		return nil
	}

	against, _ := cmd.Flags().GetString("against")
	copyFlag, _ := cmd.Flags().GetBool("copy")
	markdown, _ := cmd.Flags().GetBool("markdown")
	md, _ := cmd.Flags().GetBool("md")
	if md {
		markdown = true
	}

	// Determine base
	isStacked := false
	var base string
	if against != "" {
		if !git.BranchExists(against) {
			ui.PrintError(fmt.Sprintf("Branch '%s' does not exist.", against))
			return nil
		}
		base = against
	} else {
		p := stack.Parent(current)
		if p != "" {
			base = p
			isStacked = true
		} else {
			base = git.HeadBranch()
		}
	}

	// Gather commits
	logOut := git.RunUnchecked("log", "--oneline", "--no-decorate", base+"..HEAD")
	var commits []handoffCommit
	for _, line := range strings.Split(logOut, "\n") {
		line = strings.TrimSpace(line)
		if line == "" {
			continue
		}
		parts := strings.SplitN(line, " ", 2)
		if len(parts) == 2 {
			commits = append(commits, handoffCommit{parts[0], parts[1]})
		}
	}

	if len(commits) == 0 {
		ui.PrintInfo(fmt.Sprintf("No changes between %s and %s.", current, base))
		return nil
	}

	// File stats
	statOut := git.RunUnchecked("diff", "--stat", base+"...HEAD")
	statSummary := ""
	if statOut != "" {
		lines := strings.Split(statOut, "\n")
		statSummary = strings.TrimSpace(lines[len(lines)-1])
	}

	// Files
	filesOut := git.RunUnchecked("diff", "--name-only", base+"...HEAD")
	var files []string
	for _, f := range strings.Split(filesOut, "\n") {
		f = strings.TrimSpace(f)
		if f != "" {
			files = append(files, f)
		}
	}

	// Format
	var output string
	if markdown {
		output = formatHandoffMD(current, base, commits, statSummary, files)
	} else {
		output = formatHandoffPlain(current, base, isStacked, commits, statSummary, files)
	}

	fmt.Println(output)

	if copyFlag {
		if ui.CopyToClipboard(output) {
			fmt.Println()
			ui.PrintSuccess("Copied to clipboard.")
		} else {
			fmt.Println()
			ui.PrintWarning("Could not copy to clipboard.")
		}
	}
	return nil
}

func formatHandoffPlain(branch, base string, isStacked bool, commits []handoffCommit, stat string, files []string) string {
	var b strings.Builder
	if isStacked {
		fmt.Fprintf(&b, "%s (on %s)\n", ui.BranchStyle.Render(branch), ui.DimStyle.Render(base))
	} else {
		fmt.Fprintf(&b, "%s (vs %s)\n", ui.BranchStyle.Render(branch), ui.DimStyle.Render(base))
	}
	fmt.Fprintf(&b, "\n%s\n", ui.BoldStyle.Render(fmt.Sprintf("Commits (%d):", len(commits))))
	for _, c := range commits {
		fmt.Fprintf(&b, "  %s  %s\n", ui.HashStyle.Render(c.hash), c.msg)
	}
	// Color the stat summary line
	colored := colorStatLine(stat)
	fmt.Fprintf(&b, "\n%s\n", colored)
	fmt.Fprintf(&b, "\n%s\n", ui.BoldStyle.Render("Files:"))
	for _, f := range files {
		fmt.Fprintf(&b, "  %s\n", ui.FileStyle.Render(f))
	}
	return b.String()
}

func colorStatLine(stat string) string {
	// Color insertions green and deletions red in stat summary
	parts := strings.Split(stat, ",")
	for i, p := range parts {
		trimmed := strings.TrimSpace(p)
		if strings.Contains(trimmed, "insertion") {
			parts[i] = " " + ui.AddStyle.Render(trimmed)
		} else if strings.Contains(trimmed, "deletion") {
			parts[i] = " " + ui.DelStyle.Render(trimmed)
		}
	}
	return strings.Join(parts, ",")
}

func formatHandoffMD(branch, base string, commits []handoffCommit, stat string, files []string) string {
	var b strings.Builder
	fmt.Fprintf(&b, "## %s\n", branch)
	fmt.Fprintf(&b, "**Base:** %s · **%d commit%s** · %s\n", base, len(commits), ui.Plural(len(commits)), stat)
	fmt.Fprintf(&b, "\n### Commits\n")
	for _, c := range commits {
		fmt.Fprintf(&b, "- `%s` %s\n", c.hash, c.msg)
	}
	fmt.Fprintf(&b, "\n### Files Changed\n")
	for _, f := range files {
		fmt.Fprintf(&b, "- `%s`\n", f)
	}
	return b.String()
}

